test_files/quote_props/quote.ts

```diff
@@ -0,0 +1,23 @@
+interface Quoted {
+  [k: string]: number;
+}
+let quoted: Quoted = {};
+
+console.log(quoted.hello);
+quoted.hello = 1;
+quoted['hello'] = 1;
+
+interface QuotedMixed extends Quoted {
+  // Assume that foo should be renamed, as it is explicitly declared.
+  // It's unclear whether it's the right thing to do, user code might
```

Comment:
So the reasoning here is that if you end up with a type that has a property declaration and a string index type, and you do `foo.bar`, it should come out has `foo.bar`, but if you do `foo['baz']`, it should come out quoted. Quoting a property access when the user typed it unquoted and it has been declared somewhere seems wrong.

There's a separate question as to whether we want to disallow mixing string index types and properties in one type, as Closure does. We might want to, but that'll certainly bring some compatibility problems. It's also not generally possible without revisiting every type constructing construct in the program, as you note with the intersection type.

We should think about that, but it's a later decision.

Does that make sense?